pkg/flowcli/project/transaction.go

```diff
@@ -250,71 +250,45 @@ func (t *Transaction) AddArguments(args []cadence.Value) error {
 	return nil
 }
 
+// AddArgument add cadence typed argument
 func (t *Transaction) AddArgument(arg cadence.Value) error {
 	return t.tx.AddArgument(arg)
 }
 
-func (t *Transaction) AddAuthorizers(addresses []string) error {
-	for _, address := range addresses {
-		err := t.AddAuthorizer(address)
-		if err != nil { // return error even if one breaks
-			return err
-		}
-	}
-
-	return nil
-}
-
-func (t *Transaction) AddAuthorizer(address string) error {
-	authorizerAddress := flow.HexToAddress(address)
-	if authorizerAddress == flow.EmptyAddress {
-		return fmt.Errorf("invalid authorizer address provided %s", address)
+// AddAuthorizers add group of authorizers
+func (t *Transaction) AddAuthorizers(authorizers []flow.Address) *Transaction {
+	for _, authorizer := range authorizers {
+		t.tx.AddAuthorizer(authorizer)
 	}
 
-	t.tx.AddAuthorizer(authorizerAddress)
-	return nil
-}
-
-func (t *Transaction) SetSignerRole(role string) error {
-	t.signerRole = signerRole(role)
-
-	if t.signerRole == SignerRoleAuthorizer {
-		err := t.AddAuthorizer(t.signer.Address().String())
-		if err != nil {
-			return err
-		}
-	}
-	if t.signerRole == SignerRolePayer && t.tx.Payer != t.signer.Address() {
-		return fmt.Errorf("role specified as Payer, but Payer address also provided, and different: %s != %s", t.tx.Payer, t.signer.Address())
-	}
-
-	return nil
+	return t
 }
 
+// Sign signs transaction using signer account
 func (t *Transaction) Sign() (*Transaction, error) {
 	keyIndex := t.signer.DefaultKey().Index()
-	signerAddress := t.signer.Address()
 	signer, err := t.signer.DefaultKey().Signer(context.Background())
 	if err != nil {
 		return nil, err
 	}
 
-	if t.signerRole == SignerRoleAuthorizer || t.signerRole == SignerRoleProposer {
-		err := t.tx.SignPayload(signerAddress, keyIndex, signer)
+	if t.shouldSignEnvelope() {
+		err = t.tx.SignEnvelope(t.signer.address, keyIndex, signer)
 		if err != nil {
 			return nil, fmt.Errorf("failed to sign transaction: %s", err)
 		}
 	} else {
-		// make sure we have at least signer as authorizer else add self
-		if len(t.tx.Authorizers) == 0 {
-			t.tx.AddAuthorizer(t.signer.Address())
-		}
-
-		err := t.tx.SignEnvelope(signerAddress, keyIndex, signer)
+		err = t.tx.SignPayload(t.signer.address, keyIndex, signer)
 		if err != nil {
 			return nil, fmt.Errorf("failed to sign transaction: %s", err)
 		}
 	}
 
 	return t, nil
 }
+
+// shouldSignEnvelope checks if signer should sign envelope or payload
+func (t *Transaction) shouldSignEnvelope() bool {
```

Comment:
Nice! 👍 

Happy to discuss this more (I see the TODO), but I think we could mention the special case and link to our documentation.

e.g.

> If an account is either the transaction proposer or an authorizer in addition to being the transaction payer, they are only required to sign the transaction envelope.

Reply:
I will add the link to the comment yes. I marked this to double-check with you validity of this implementation. Docs say if it's payer and either proposer or authorizer it signs envelope, but in my implementation this could be reduced logically to if it is a payer sign envelope. Wanted to make sure it's ok. In testing works.